WeBWorK: tikz image

Alex-Jordan opened the pull request:
This lets you author `image/latex-image[@syntax="PGtikz"]` inside a WeBWorK exercise.

* Examples in the sample chapter, and one additional example in the showcase article.
* Schema
* Guide
* `pretext` script might be downloading a .tgz file if the problem has tikz code in it. This contains the tex, pdf, svg, and png that the WW server built. The script unpacks and keeps them all. Note this is needed even if it is a server problem that is using tikz (as opposed to a PTX-authored problem)

Nothing interferes with the building of other "latex-image" elements by the `pretext` script? Why not? Because by the time it is searching the tree for `image[latex-image]`, all the `webwork//image[latex-image/@syntax="PGtikz"]` have become `webwork-reps//image[@source]`, so they are not found by the usual latex-image extraction.

First thing to do with testing is build sample chapter (graphics section) and showcase article (webwork section) and see that the images in WW are working. Build both HTML and PDF.

Alex-Jordan commented:
Something I meant to add. We will also want these images to respect whatever is in "latex-image-preamble". (Or possibly a special "latex-image-preamble" that only applies to `image/latex-image[@syntax="PGtikz"]`.) That will happen with #1470, which will be cleaned up following this PR.

We don't want to directly drop "latex-image-preamble" into a problem file. It can make the problem file so long that its base64 encoding could surpass the cap on URL length. So "latex-image-preamble" needs to end up in a separate WW macro library file that you build one time and place in the host course. (Well, you build it one time, but maybe every now and then you need to update it.)

rbeezer commented:
Really, really nice.  Comments and nits, in no particular order.

*  I've been reserving `CDATA` for really big messy stuff.  For really simple stuff I've been writing bare `&lt;`, etc.
*  And I'm not a fan of `&#x2d;` in final products.  ;-)  (Yes, I know why you do it!)
*  "You may enter the answer using C(n,k) notation."  "may" or "must"?
*  With `-v` we get `['image.tex', 'image.pdf', 'image.svg', 'image.png']`.  Does not seem so helpful as a progess indicator.  Should it be `-vv`?
*  An example with an x-y curve and axes, etc., would be really helpful for authors, I'd think.  That can definitely wait for later.
*  Do you want to trash the `*.tgz` after it is unpacked?
*  With `xelatex` on the Showcase, I get
```
[11GPL Ghostscript 9.50: Unrecoverable error, exit code 1
GPL Ghostscript 9.50: Unrecoverable error, exit code 1
GPL Ghostscript 9.50: Unrecoverable error, exit code 1
GPL Ghostscript 9.50: Unrecoverable error, exit code 1
GPL Ghostscript 9.50: Unrecoverable error, exit code 1
GPL Ghostscript 9.50: Unrecoverable error, exit code 1
GPL Ghostscript 9.50: Unrecoverable error, exit code 1
] [12]
```
The Venn digaram, in various formats, all look good, but in the PDF it is degraded.  Is this the PSTricks issue we were working on before?

Alex-Jordan commented:
Force pushed with updates.

> I've been reserving CDATA for really big messy stuff. For really simple stuff I've been writing bare &lt;, etc.
OK, changed to `&lt;`.

> And I'm not a fan of &#x2d; in final products. ;-) (Yes, I know why you do it!)
The more I do this, the more I stop needing to look up that it's `&#x2d;`. And I do find it handy to comment out chunks of the sample chapter during development. So unless you insist, I would like to leave that.

> "You may enter the answer using C(n,k) notation." "may" or "must"?

Well, it was "must", but only because imho there is a bug in the underlying WW macro library that allows using `C(n,k)`. Really I think that should be fixes so that `C(n+1,2)` and `(n+1)n/2` are equivalent. But in the meantime I modified the code to explicitly take either option.

> With -v we get ['image.tex', 'image.pdf', 'image.svg', 'image.png']. Does not seem so helpful as a progess indicator. Should it be -vv?

Turns out that was a print statement accidentally left in place from development debugging. Gone now.

> An example with an x-y curve and axes, etc., would be really helpful for authors, I'd think. That can definitely wait for later.

I think that is the perfect thing to go with a "latex-image-preamble" that styles the axes. So it's next up in the PR queue.

> Do you want to trash the *.tgz after it is unpacked?

It's trashed now. This did get me thinking that we should automatically package up all image files to make it easier to send them to, say, a disability services office.

> With xelatex on the Showcase, I get

For me with TeXLive 2021, I get no issues there with xelatex. However, to build PDF of the showcase article, I had to comment out the pstricks images. As mentioned in some other thread, now with TeXLive 2021 you need a special driver argument for xelatex (and I think for pdflatex too?) if you want to use those tools on a file with a pstricks image. So there is a larger issue. Someone with an up to date TeXLive installation can't build the Showcase article without knowing that special thing. Can't build it as PDF, and can't build those images as SVG for HTML either.

With the force push, I also committed the new image files to the showcase images folder.

rbeezer commented:
No luck solving the xelatex/GhostScript error.  Not

`-output-driver="xdvipdfmx -i dvipdfmx-unsafe.cfg -q -E"`

nor

https://bbs.archlinux.org/viewtopic.php?id=250323

The Showcase Article will just have to suffer, which is contrary to its purpose.  Back to the meat of the PR...

rbeezer commented:
> So unless you insist, I would like to leave that.

Not a request - just opinion.

> Turns out that was a print statement accidentally left in place

Thought so!  Did you like my diplomatic query?  ;-)

> So it's next up in the PR queue.

Perfect.

>  we should automatically package up all image files

Maybe when image management is done we can have a component/format combination for the `pretext/pretext` script that will do this.

Alex-Jordan commented:
> No luck solving the xelatex/GhostScript error.

Not to derail....but are you still using TeXLive 2019? I think the first step to see the issue and then think about the global fix is to move to 2021.

rbeezer commented:
* Venn diagram now looks better, and I think only change is using the version comitted in the repo.  There is still a GS error though...  Hmmmm.
* Venn diagram has archive link for EPS, no such file exists.
* Do we have to have files comitted with filenames like: `webwork-12-image-1.pdf`?
* Merged, pushed, website rebuilding just now.  I'm unlikely to have time to check the website versions right away.

This making the WW look really nice.  We'll look back on the iframe versions and shake our heads.  ;-)

rbeezer commented:
Ubuntu 20.04 LTS, which will next do a major upgrade Spring 2022.  Not sure I can easily move off of TL 2019 now.

Alex-Jordan commented:
Thanks, and the credit for this really goes to Glenn Rice for getting WW to handle tikz. You'll meet him end of next month.

> Venn diagram has archive link for EPS, no such file exists.

Yeah... I knew about that. Not sure what to do. Maybe manually make an .eps version. I'd prefer to turn off eps archive at that image, but that would be a project since we don't track that attribute through the webwork conversion process. Somehow it's not an issue for the non-tikz image, so maybe I can look into that and make more sense of things.

> Do we have to have files comitted with filenames like: webwork-12-image-1.pdf?

I think yes. The image is born in an "exercise/webwork//image/latex-image". So in principle you could give the "image" a name or id. But then we ship it to the server as part of a .pg file where there is no mechanism to include that name or id. Then we grab the image files form the server. So we would have to be aligning the PTX source image (with its name or id) with the PTX static that we get back from the WW server. I've tried that before (for example to recognize when an "ol" was authored with "@cols") and it doesn't go well. I mean, it can work, but the ratio of overhead code for it raises sustainability and fragility issues.

>  We'll look back on the iframe versions and shake our heads. ;-)

Well thanks for the nod to my disapproval of iframes. It turns out we will have to bring back iframes for _some_ problems though. Namely the ones that have their own javascript doohickey in them. It's OK though because we will be able to know when that happens and the div injection can just be an iframe injection instead. So not PTX XSL work, but rather pretext-webwork.js work. And I believe Glenn can help me make that happen.

Alex-Jordan commented:
> Merged, pushed, website rebuilding just now. I'm unlikely to have time to check the website versions right away.

Only thing I notice is that the new image files might need to be uploaded the server's images folder.

rbeezer commented:
I could benefit from a **loud** reminder to update derived schema stuff when main version gets an edit.  ;-)

Alex-Jordan commented:
Got it. (Was something wrong with that?)

On Thu, May 20, 2021 at 12:52 PM Rob Beezer ***@***.***>
wrote:

> I could benefit from a *loud* reminder to update derived schema stuff
> when main version gets an edit. ;-)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rbeezer/mathbook/pull/1476#issuecomment-845429694>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABEDOAARZRBFTIEP2YFIZCLTOVR7BANCNFSM45BS6ORA>
> .
>


-- 
Alex Jordan
Mathematics Instructor
Portland Community College

rbeezer commented:
No nothing wrong at all.  I have a script to keep three other files up-to-date.  I like to have those changes in the vicinity of the (source) changes in `[retext.xml`.  I forgot this time.  I just need **help** reminding me what needs to be done.

rbeezer commented:
`s/[retext.xml/pretext.xml/`

Alex-Jordan commented:
OK, just checking. I was uncertain about the mucking about I did in
pretext.xml with the WWImage item. I would not have been surprised if
it had messed something up.

On Thu, May 20, 2021 at 1:30 PM Rob Beezer ***@***.***> wrote:
>
> s/[retext.xml/pretext.xml/
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.



-- 
Alex Jordan
Mathematics Instructor
Portland Community College

rbeezer commented:
> Only thing I notice is that the new image files might need to be uploaded the server's images folder.

Image upload got fixed as part of #1478

rbeezer commented:
See 8e6285add032790816f06b3d68c9afb91719f761.
